### ISSUE TITLE
Overlapping overload of __get__ method

### ISSUE THREAD

**Issue by ermishechkin** [microsoft/pyright#1719]
**Describe the bug**
I am trying to annotate a class that acts as a property when used with a specific container class and as a regular object elsewhere. According to https://github.com/python/typing/issues/253#issuecomment-389262904 it is allowed to specify different return types, even for overlapping overloads. But pyright emits ```Overload 2 for "__get__" overlaps overload 3 and returns an incompatible type``` error for the following code (since 1.1.118).

There is also another error (since 1.1.117) that can be a duplicate of https://github.com/microsoft/pyright/issues/1718:
```
Overloaded function implementation is not consistent with signature of overload 1
  Type "(self: Descriptor, owner: Any | None, inst: Type[Any]) -> (Descriptor | int)" cannot be assigned to type "(self: Descriptor, owner: None, inst: Type[Any]) -> Descriptor"
    Parameter 3: type "Type[Any]" cannot be assigned to type "Type[Any]"PylancereportGeneralTypeIssues
```

**Screenshots or Code**
```python
from __future__ import annotations
from typing import Any, Literal, Optional, Type, TypeVar, Union, overload

TContainer = TypeVar('TContainer')

class Descriptor:
    @overload
    def __get__(self, owner: Literal[None], inst: Type[Any]) -> Descriptor:
        ...

    @overload
    def __get__(self, owner: Container, inst: Type[Container]) -> int:
        ...

    @overload
    def __get__(self, owner: TContainer, inst: Type[TContainer]) -> Descriptor:
        ...

    def __get__(self, owner: Optional[Any],
                inst: Type[Any]) -> Union[Descriptor, int]:
        if isinstance(owner, Container):
            return 7  # some special case
        return self


class Container:
    field = Descriptor()


class UnknownContainer:
    field = Descriptor()


reveal_type(Container.field)            # Type is "Descriptor"
reveal_type(Container().field)          # Type is "int"
reveal_type(UnknownContainer.field)     # Type is "Descriptor"
reveal_type(UnknownContainer().field)   # Type is "Descriptor"
```

**VS Code extension or command-line**
pyright 1.1.127

**Comment by erictraut** [comment]
It is odd that the `__get__` method would be exempt from this check. The mypy documentation that you referenced doesn't provide any explanation for why this is the case. But in the interest of following precedent, I'll add the same special case in pyright. This change will be included in the next release.

**Comment by erictraut** [comment]
This fix is included in pyright 1.1.128, which I just published. It will also be included in the next release of pylance.
